modules/saas.py

```diff
@@ -0,0 +1,2542 @@
+        self.url = f"{url}/api/v{self.version}"
+        self.verify = verify
+        self._access = None
+
+        self.token_headers = {
+            "Content-Type": "application/x-www-form-urlencoded",
+        }
+
+        if client_id and secret_key:
+            self.auth = (client_id, secret_key)
+        else:
+            raise RuntimeError(
+                "API Client ID and Secret Key are required for authentication."
+            )
+
+    @staticmethod
+    def _remove_trailing_slashes(url):
+        if ":/" not in url:
+            url = "https://" + url
+        else:
+            url = re.sub("^.*://?", "https://", url)
+        url = url[:-1] if url.endswith("/") else url
+        return url
+
+    @request_error_handler
+    def _request(self, method, url, **kwargs):
+        """
+        Do a get request on the provided URL
+        This is used by paginated endpoints
+        :rtype: requests.Response
+        """
+        self._check_token()
+        if method not in ["get", "patch", "put", "post", "delete"]:
+            raise ValueError("Invalid requests method provided")
+
+        if "headers" in kwargs:
+            headers = kwargs.pop("headers")
+        else:
+            headers = {
+                "Authorization": f"Bearer {self._access}",
+                "Content-Type": "application/json",
+            }
+        return requests.request(
+            method=method, url=url, headers=headers, verify=self.verify, **kwargs
+        )
+
+    def _sleep(self, timeout):
+        time.sleep(timeout)
+
+    def _refresh_token(self):
+        if self._refreshTime > int(time.time()):
+            data = {"grant_type": "refresh_token", "refresh_token": self._refresh}
+            response = requests.post(
+                url=f"{self.base_url}/oauth2/token",
+                headers=self.token_headers,
+                auth=self.auth,
+                data=data,
+            )
+            token_data = response.json()
+            self._access = token_data["access_token"]
+            self._accessTime = int(time.time()) + token_data["expires_in"] - 100
+        else:
+            self._get_token()
+
+    def _get_token(self):
+        data = {"grant_type": "client_credentials"}
+        response = requests.post(
+            url=f"{self.base_url}/oauth2/token",
+            headers=self.token_headers,
+            auth=self.auth,
+            data=data,
+        )
+        token_data = response.json()
+        self._access = token_data["access_token"]
+        self._refresh = token_data["refresh_token"]
+        self._accessTime = int(time.time()) + token_data["expires_in"] - 100
+        self._refreshTime = int(time.time()) + token_data["refresh_expires_in"] - 100
+
+    def _check_token(self):
+        if not self._access:
+            self._get_token()
+        if self._accessTime < int(time.time()):
+            self._refresh_token()
+
+    @staticmethod
+    def _generate_detection_params(args):
```

Comment:
Are these a copy and paste from the methods in vectra.py?  We should be able to use inheritance if so to avoid DRY violations here.

If we inherit from vectra.VectraClient and override the __init__(), would that cut down on a lot of the code changes needed here?